### DISCUSSION TITLE
map decoded signals

### DISCUSSION THREAD

**Discussion by nielsgisbertz** [bengtmartensson/Infrared4Arduino#69] [Q&A]
How do I use the decoded signals in a case structere? I tried to use the stored integeres from the library but without succes.

**Comment by bengtmartensson** [comment]
Not quite sure what you are asking for, but I guess it is something like this:

```
void loop() {
    receiver->receive();

    if (receiver->isEmpty())
        return; // timeout

    Nec1Decoder decoder(*receiver);
    if (decoder) {
           if (decoder.getD() != 12 || decoder.getS() != 34) // check device & subdevice (optional)
               return;
           switch (decoder.getF()) {
                case 0:
                     zero();
                     break;
                case 1:
                     one();
                     break;
               // ....
               default:
                    unknown();
           }
    } else {
         // not a NEC
    }
}

```

**Reply by nielsgisbertz** [reply]
This is indeed what I ment and what I tried.
think I'm close but since I did not managed to analyze the format of the stored intergers I'm still guessing.
I retrieved the case 2, 3, 9 and 0 from the Nec1Decoder:
NEC1 4 2 -> case 2
NEC1 4 3 -> case 3
NEC1 4 9 -> case 9

```
void loop() {

    int times = 1; 
    receiver->receive();

    if (receiver->isEmpty())
        Serial.println(F("timeout"));


    Nec1Decoder decoder(*receiver);
                decoder.printDecode(Serial);
    
    if (decoder) 
           {
           if (decoder.getD() != 12 || decoder.getS() != 34) // check device & subdevice (optional)
               return;
   
              
           switch (decoder.getF()) {
                case 2:
                     sendRc5(16U, 16U, times); // vol up
                     Serial.print("Vol up");
                     break;
                case 3:
                     sendRc5(16U, 17U, times); // vol down
                     Serial.print("Vol down");
                     break;
                case 9:
                     sendRc5(16U, 12U, times); //mute
                     Serial.print("mute");
                case 0:
                     digitalWrite(12, HIGH); // sets the digital pin 13 on
                     delay(100);             // waits for 100ms (TBD)
                     digitalWrite(12, LOW);  // sets the digital pin 13 off
                     Serial.println("Source Select");
                   
               default:
                   // delay(10);   //tbd

                    }
    } 
    
    
    else {
         
         Serial.println(F("No NEC code recieved"));// not a NEC
    }
    
}
```

**Reply by bengtmartensson** [reply]
Not sure if/where there is still a problem. Do you know how the line

    if (decoder.getD() != 12 || decoder.getS() != 34)

works? Also don't forget the "break"s.

**Reply by nielsgisbertz** [reply]
No, I don't know how that line works. but normaly "!" is inverted.
Quit new to programming and somtimes I struggle to find the correct search terms to lookup the syntax and code structure.
Thanks already fro the help!

Ah, I see I've indeed forgotten some breaks
Will give that try. and let you know
edit: 
The program prints the code it received, and doesn't loops each time trough the "default" and "else" structure.
I only don't get any output from the case structure.


FYI: 
Goal is to use the LG remote from my TV to control some basics of my Marantz PM7000.
The source select is to give a pulse to my Raspberry Streamer/DAC to switch sources. This is not tested yet because I'm waiting for the extra board from Audiophonics.

**Reply by bengtmartensson** [reply]
A NEC decode is characterized by three (8-bit) parameters D, S and F. Normally, you require all three to match in order to trigger an action. The if (...) line  checks that the D and the S parameters are 12 and 34 respectively.  If they do not match, no further processing is made. This was intended as an example. It appears that you should change 12 to 4 and 34 to 251 (=255 - 4).

Hope this helps!

**Reply by nielsgisbertz** [reply]
Now I get a output on the led. 
But no reaction from the PM 7000

these are the values from the Multidecode example.
The 1/0 is the toggle for each sequential button press


// Command #1: MAIN VOLUME UP
// Protocol: rc5, Parameters: D=16U F=16U (output on the Serial Monitor: RC5 16 16 1/0)

// Command #2: MAIN VOLUME DOWN
// Protocol: rc5, Parameters: D=16U F=17U (output on the Serial Monitor: RC5 16 17 1/0)

// Command #4: MUTE
// Protocol: rc5, Parameters: D=16U F=13U (output on the Serial Monitor: RC5 16 13 1/0)


Led blinks: check bij Phone camera, visible red led in series  and has reacted, by chance,  once on a IRRemote program.

**Reply by bengtmartensson** [reply]
Seems that the original problem is solved.

> But no reaction from the PM 7000

Sorry not enough information. Possibly the signal parameters are wrong?

**Reply by nielsgisbertz** [reply]
I'll set up some experiments and come back to you.
Let's finish this, then document it and use it as example to help others to use your library.
